Updating the warnings for Fusion (#8981)
fishtown-analytics/docs.getdbt.com: website/docs/reference/global-configs/warnings.md

`WARN_ERROR` and `WARN_ERROR_OPTIONS` are mutually exclusive. You can only specify one, even when you're specifying the config in multiple places (like env var or a flag), otherwise, you'll see a usage error.
:::

Warnings that should be treated as errors can be specified through the `error` parameter. Warning names can be found in:
  - [dbt-core's types.py file](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/events/types.py), where each class name that inherits from `WarnLevel` corresponds to a warning name (e.g. `AdapterDeprecationWarning`, `NoNodesForSelectionCriteria`).
  - Using the `--log-format json` flag.

<VersionBlock firstVersion="2.0">

> In the <Constant name="fusion_engine" />, every warning has both a numeric code (for example, `1092`) and an event name (for example, `NoNodesForSelectionCriteria`). You can use either form interchangeably in `warn_error_options`. Messages printed at runtime include both the name and the code. See [Supported legacy dbt-Core event name aliases](#supported-legacy-dbt-core-event-name-aliases) for the full list of supported event names and their corresponding codes.

Reviewer: Elias said they wanted us to be opinionated and only support names (even though we print codes in the message for our own automation). We haven’t yet decided whether we’ll keep capability to specify numeric codes, but I think everyone aligned that even if we do, it should not be documented

> In the <Constant name="fusion_engine" />, warning codes are numeric (for example, `1600`, `1074`) rather than the event class names used in <Constant name="core" />. <Constant name="fusion" /> uses numeric codes as the canonical configuration key, with a subset of legacy <Constant name="core" /> event names supported as aliases. You can find <Constant name="fusion" /> warning codes using the `--log-format json` flag, which includes the numeric code in the `code` field of each log entry. See [Supported legacy dbt-Core event name aliases](#supported-legacy-dbt-core-event-name-aliases) for the full list of supported event name aliases.

Reviewer: > You can find <Constant name="fusion" /> warning codes using the `--log-format json` flag

codes are also present in warning messages directly in the form of `dbtNNNN`

<img width="2018" height="198" alt="Image" src="https://github.com/user-attachments/assets/8d3a69da-66dd-4148-8bcd-f2e5f64225e5" />

> In the <Constant name="fusion_engine" />, warning codes are numeric (for example, `1600`, `1074`) rather than the event class names used in <Constant name="core" />. <Constant name="fusion" /> uses numeric codes as the canonical configuration key, with a subset of legacy <Constant name="core" /> event names supported as aliases. You can find <Constant name="fusion" /> warning codes using the `--log-format json` flag, which includes the numeric code in the `code` field of each log entry. See [Supported legacy dbt-Core event name aliases](#supported-legacy-dbt-core-event-name-aliases) for the full list of supported event name aliases.

Author: ```suggestion
In the <Constant name="fusion_engine" />, warning codes are numeric (for example, `1600`, `1074`) rather than the event class names used in <Constant name="core" />. <Constant name="fusion" /> uses numeric codes as the canonical configuration key, with a subset of legacy <Constant name="core" /> event names supported as aliases. The <Constant name="fusion" /> warning codes will be present in any warnings displayed and you can also find them using the `--log-format json` flag, which includes the numeric code in the `code` field of each log entry. See [Supported legacy dbt-Core event name aliases](#supported-legacy-dbt-core-event-name-aliases) for the full list of supported event name aliases.
```

> In the <Constant name="fusion_engine" />, warning codes are numeric (for example, `1600`, `1074`) rather than the event class names used in <Constant name="core" />. <Constant name="fusion" /> uses numeric codes as the canonical configuration key, with a subset of legacy <Constant name="core" /> event names supported as aliases. The <Constant name="fusion" /> warning codes will be present in any warnings displayed and you can also find them using the `--log-format json` flag, which includes the numeric code in the `code` field of each log entry. See [Supported legacy dbt-Core event name aliases](#supported-legacy-dbt-core-event-name-aliases) for the full list of supported event name aliases.

Reviewer 2: ```suggestion
In the <Constant name="fusion_engine" />, warning codes are numeric (for example, `1600`, `1074`) rather than the event class names used in <Constant name="core" />. <Constant name="fusion" /> uses numeric codes as the canonical configuration key, with a subset of legacy <Constant name="core" /> event names supported as aliases. The <Constant name="fusion" /> warning codes will be present in any warnings displayed and you can also find them using the `--log-format json` flag, which includes the numeric code in the `code` field of each log entry. Refer to [Supported legacy dbt-Core event name aliases](#supported-legacy-dbt-core-event-name-aliases) for the full list of supported event name aliases.
```

Reviewer 2: Hiya Matt

Should this reference to "dbt-Core" be "dbt-Core", "dbt-core" like it is below in the H3 header or have a constant variable?

Reviewer 2: And also maybe start the sentence in second person so that it adheres to the style guide "You can identify <Constant name="fusion" />  warning codes..."

Reviewer 2: I might have missed it, but is that 1600 warn / error correct? I couldn't find it in the repo and the doc examples 1601 as the warn / error

This means that if a new warning is introduced in a future version of <Constant name="core" />, your production job may start failing unexpectedly. We recommend proceeding with caution when doing this in production environments, and explicitly listing only the warnings you want to treat as errors in production.
:::

<VersionBlock firstVersion="2.0">

## Fusion behavior and warning codes

The <Constant name="fusion_engine" /> fully supports `warn_error_options`. This section describes important differences from <Constant name="core" /> behavior.

Existing dbt-core event names fall into three categories:

> - **Supported** Mapped to similar <Constant name="fusion" /> warning and behave approximately the same.

Reviewer 2: ```suggestion
- **Supported:** Mapped to similar <Constant name="fusion" /> warning and behave approximately the same.
```

- **Won't be supported:** Those that we deliberately decided to not ever support.

> - **Not supported yet:** Parsed, but do nothing yet.

Reviewer 2: This is a non-block suggestion (as I can see the intention ). Bold is usually used for UI elements, maybe remove the bold or add the info to a small table?

flags:
  warn_error_options:
    error:

>       - 1600   # NoNodesForSelectionCriteria

Reviewer: can we use 1601 (`NothingToDo`) since `NoNodesForSelectionCriteria` is not really supported today (but soon)

Author: Good call out

>       - 1600   # NoNodesForSelectionCriteria

Author: ```suggestion
      - 1601   # NothingToDo
```

      - 1074   # JinjaLogWarning / exceptions.warn()
    silence:
      - 1078   # FreshnessConfigProblem
```

> A subset of legacy <Constant name="core" /> event names are also accepted as aliases and mapped to their corresponding <Constant name="fusion" /> codes. Any value that is not a recognized numeric code, supported legacy event name, or supported group (`all`, `*`) causes <Constant name="fusion" /> to exit with an error at startup. This is a deliberate change from <Constant name="core" />, which silently ignored unknown values.

Reviewer: > This is a deliberate change from <Constant name="core" />, which silently ignored unknown values.

this is not true. dbt-core also fails on invalid values

Reviewer: dbt-core also fails when `warn` key is used without `error: "all"` or `--warn-error` - but fusion happily allows this as of today ;-)

      - 1074   # JinjaLogWarning / exceptions.warn()
    silence:
      - 1078   # FreshnessConfigProblem
```

A subset of legacy <Constant name="core" /> event names are also accepted as aliases and mapped to their corresponding <Constant name="fusion" /> codes. Any value that is not a recognized numeric code, supported legacy event name, or supported group (`all`, `*`) causes <Constant name="fusion" /> to exit with an error at startup.

> ### Supported legacy dbt-core event name aliases

Reviewer 2: Same as above, should this be "dbt-core" or "dbt-Core"?

> A subset of legacy <Constant name="core" /> event names are also accepted as aliases and mapped to their corresponding <Constant name="fusion" /> codes. Any value that is not a recognized numeric code, supported legacy event name, or supported group (`all`, `*`) causes <Constant name="fusion" /> to exit with an error at startup. This is a deliberate change from <Constant name="core" />, which silently ignored unknown values.

Author: ```suggestion
A subset of legacy <Constant name="core" /> event names are also accepted as aliases and mapped to their corresponding <Constant name="fusion" /> codes. Any value that is not a recognized numeric code, supported legacy event name, or supported group (`all`, `*`) causes <Constant name="fusion" /> to exit with an error at startup.
```

> | 1600 | `NoNodesForSelectionCriteria` | `--select` criteria matched no nodes |

Reviewer: not yet

> | 1600 | `NoNodesForSelectionCriteria` | `--select` criteria matched no nodes |

Author: ```suggestion
```

| 1601 | `NothingToDo` | No nodes selected (alias) |
| 1087 | `NodeNotFoundOrDisabled` | A test or exposure dependency references a missing or disabled node |
| 1085 | `DeprecatedModel` | A model has passed its deprecation date and should be removed |
| 1072 | `DeprecatedReference` | A reference to a model that has already been deprecated |
| 1073 | `UpcomingReferenceDeprecation` | A reference to a model that will be deprecated on a future date |
| 1074 | `JinjaLogWarning` | Jinja `exceptions.warn()` called in a macro |
| 1075 | `SnapshotTimestampWarning` | Snapshot timestamp column type mismatch |
| 1076 | `PackageRedirectDeprecation` | A package has been redirected to a new name; update your `packages.yml` |
| 1077 | `DepsUnpinned` | A git-sourced package uses an unpinned revision (`HEAD`, `main`, or `master`) |
| 1078 | `FreshnessConfigProblem` | A source has no freshness configuration; freshness check was skipped |
| 1084 | `WarnStateTargetEqual` | The `--state` and `--target` directories are the same path |
| 1086 | `WEOIncludeExcludeDeprecation` | Deprecated `include`/`exclude` keys were used in `warn_error_options`; use `error`/`warn` instead |

> | _(no code)_ | `LogTestResult` | A data test result (pass/warn/fail); matched by name only |

Author: ```suggestion
| _no code_ | `LogTestResult` | A data test result (pass/warn/fail); matched by name only |
```

| 1077 | `DepsUnpinned` | A git-sourced package uses an unpinned revision (`HEAD`, `main`, or `master`) |
| 1078 | `FreshnessConfigProblem` | A source has no freshness configuration; freshness check was skipped |
| 1084 | `WarnStateTargetEqual` | The `--state` and `--target` directories are the same path |
| 1086 | `WEOIncludeExcludeDeprecation` | Deprecated `include`/`exclude` keys were used in `warn_error_options`; use `error`/`warn` instead |
| _no code_ | `LogTestResult` | A data test result (pass/warn/fail); matched by name only |

> ### Unsupported Core event names

Reviewer 2: ```suggestion
### Unsupported dbt Core event names
```

> Only the legacy names in [Supported legacy dbt-Core event name aliases](#supported-legacy-dbt-core-event-name-aliases) are valid string aliases in <Constant name="fusion" />. There are many other <Constant name="core" /> warning event names; if you put one of those in `warn_error_options`, <Constant name="fusion" /> fails at startup, as described in [Warning codes in Fusion](#warning-codes-in-fusion).

Reviewer: no, that's not the case. fusion will warn if any of the 263 event names that dbt-core has but not among the supported list are used. Not fail. It will only fail on names that are not one of the true dbt-core event names. So it is parsing almost exactly as dbt-core, but warns when legacy events are not mapped to fusion codes internally yet

> Only the legacy names in [Supported legacy dbt-Core event name aliases](#supported-legacy-dbt-core-event-name-aliases) are valid string aliases in <Constant name="fusion" />. There are many other <Constant name="core" /> warning event names; if you put one of those in `warn_error_options`, <Constant name="fusion" /> will throw a warning at startup.

Reviewer 2: Same as above, should this be "dbt-core" or "dbt-Core"?

> The table below is not a complete list of unsupported names. It only includes <Constant name="core" /> event names that <Constant name="fusion" /> recognizes by name so it can emit a startup warning explaining why the entry has no effect and prompting you to remove it: the underlying <Constant name="core" /> behavior was removed, replaced, or made unconditional in <Constant name="fusion" />. Many other unsupported <Constant name="core" /> names are not listed here; they still fail startup validation when used in `warn_error_options`.

Reviewer: this list may eventually be generated from fusion to ensure it always exactly matches in the docs. Just to re-iterate - all known (264) dbt-core event names fall under one of the 3 categories:
* supported, aka mapped to similar fusion warning and behave roughly the same
* will not be supported, those that we deliberately decided to not ever support. The list below. May grow, although I doubt it will grow significantly
* not supported yet - parsed, but do nothing yet. to be supported
  * one caveat here - in reality out of these most are not warnings in dbt-core, so there is nothing to support. but we haven't analyzed this, so we just assume any dbt-core event may be a warning

The table below is not a complete list of unsupported names. It only includes <Constant name="core" /> event names that <Constant name="fusion" /> recognizes by name so it can emit a startup warning explaining why the entry has no effect and prompting you to remove it: the underlying <Constant name="core" /> behavior was removed, replaced, or made unconditional in <Constant name="fusion" />. Many other unsupported <Constant name="core" /> names are not listed here; they still warn during startup validation when used in `warn_error_options`.

> | dbt-Core event name | Reason not supported |

Author: ```suggestion
| dbt-core event name | Message |
```

|---|---|

> | `MicrobatchMacroOutsideOfBatchesDeprecation` | This warning comes from partial parsing in dbt Core, which Fusion does not support. |

Reviewer 2: The table needs constant variables for fusion and core

> | `MicrobatchMacroOutsideOfBatchesDeprecation` | This warning comes from partial parsing in dbt Core, which Fusion does not support. |

Reviewer 2: ```suggestion
| `MicrobatchMacroOutsideOfBatchesDeprecation` | This warning comes from partial parsing in <Constant name="core" />, which <Constant name="fusion" /> does not support. |
```

| `SeedExceedsLimitSamePath` | This warning comes from partial parsing in dbt Core, which Fusion does not support. |
| `SeedIncreased` | This warning comes from partial parsing in dbt Core, which Fusion does not support. |
| `GenerateSchemaNameNullValueDeprecation` | Fusion only supports the newer behavior-change flag, where this case is a hard error. |

> | `GenericSemanticLayerDeprecation` | Fusion already implements the new semantic layer spec, so this legacy warning no longer applies. |

Reviewer 2: ```suggestion
| `GenericSemanticLayerDeprecation` | Fusion already implements the new <Constant name="semantic_layer" /> spec, so this legacy warning no longer applies. |
```

> | `MFCumulativeTypeParamsDeprecation` | Fusion already implements the new semantic layer spec, so this legacy warning no longer applies. |

Reviewer 2: ```suggestion
| `MFCumulativeTypeParamsDeprecation` | Fusion already implements the new <Constant name="semantic_layer" /> spec, so this legacy warning no longer applies. |
```

> | `MFTimespineWithoutYamlConfigurationDeprecation` | Fusion already implements the new semantic layer spec, so this legacy warning no longer applies. |

Reviewer 2: ```suggestion
| `MFTimespineWithoutYamlConfigurationDeprecation` | Fusion already implements the new <Constant name="semantic_layer" /> spec, so this legacy warning no longer applies. |
```

> | `MetricAttributesRenamed` | Fusion already implements the new semantic layer spec, so this legacy warning no longer applies. |

Reviewer 2: ```suggestion
| `MetricAttributesRenamed` | Fusion already implements the new <Constant name="semantic_layer" /> spec, so this legacy warning no longer applies. |
```

> | `TimeDimensionsRequireGranularityDeprecation` | Fusion already implements the new semantic layer spec, so this legacy warning no longer applies. |

Reviewer 2: ```suggestion
| `TimeDimensionsRequireGranularityDeprecation` | Fusion already implements the new <Constant name="semantic_layer" /> spec, so this legacy warning no longer applies. |
```

> | `WrongResourceSchemaFile` | Fusion emits `NoNodeForYamlKey` (dbt1005) instead, due to parallelism at parse time |

Reviewer: I haven't verified this list, but here is the source of truth from the [code](https://github.com/dbt-labs/fs/blob/782624920dcc6440aa0875143ea5b7225a7ef981/fs/sa/crates/dbt-common/src/warn_error_options/legacy.rs#L313-L390):

```rust
pub enum WillNotSupportLegacyWarnError {
    #[strum(
        message = "Fusion only supports the newer behavior-change flag, where this case is a hard error."
    )]
    MicrobatchMacroOutsideOfBatchesDeprecation,
    #[strum(
        message = "This warning comes from partial parsing in dbt Core, which Fusion does not support."
    )]
    SeedExceedsLimitSamePath,
    #[strum(
        message = "This warning comes from partial parsing in dbt Core, which Fusion does not support."
    )]
    SeedIncreased,
    #[strum(
        message = "Fusion only supports the newer behavior-change flag, where this case is a hard error."
    )]
    GenerateSchemaNameNullValueDeprecation,
    #[strum(
        message = "Fusion already implements the new semantic layer spec, so this legacy warning no longer applies."
    )]
    GenericSemanticLayerDeprecation,
    #[strum(
        message = "Fusion already implements the new semantic layer spec, so this legacy warning no longer applies."
    )]
    MFCumulativeTypeParamsDeprecation,
    #[strum(
        message = "Fusion already implements the new semantic layer spec, so this legacy warning no longer applies."
    )]
    MFTimespineWithoutYamlConfigurationDeprecation,
    #[strum(
        message = "Fusion already implements the new semantic layer spec, so this legacy warning no longer applies."
    )]
    MetricAttributesRenamed,
    #[strum(
        message = "Fusion already implements the new semantic layer spec, so this legacy warning no longer applies."
    )]
    TimeDimensionsRequireGranularityDeprecation,
    #[strum(
        message = "Fusion already uses the newer source freshness behavior, so this legacy warning does not apply."
    )]
    SourceFreshnessProjectHooksNotRun,
    #[strum(message = "Fusion does not support semantic models, so this warning does not apply.")]
    SemanticValidationFailure,
    #[strum(
        message = "Fusion already validates allowed YAML keys strictly, so this warning would be redundant."
    )]
    ValidationWarning,
    #[strum(
        message = "Fusion already enforces the latest behavior, which prevents packages from overriding built-in materializations."
    )]
    PackageMaterializationOverrideDeprecation,
    #[strum(
        message = "Fusion does not surface this warning by default, which matches current dbt Core behavior."
    )]
    TestsConfigDeprecation,
    #[strum(
        message = "Fusion already errors on this configuration, which matches newer dbt Core behavior."
    )]
    ProjectFlagsMovedDeprecation,
    #[strum(message = "This is now fully deprecated in Fusion.")]
    ConfigSourcePathDeprecation,
    #[strum(message = "This is now fully deprecated in Fusion.")]
    ConfigLogPathDeprecation,
    #[strum(message = "This is now fully deprecated in Fusion.")]
    ConfigTargetPathDeprecation,
    #[strum(message = "This is now fully deprecated in Fusion.")]
    ConfigDataPathDeprecation,
    #[strum(
        message = "Fusion reserves the DBT_ENGINE_ prefix and rejects unknown environment variables that use it."
    )]
    EnvironmentVariableNamespaceDeprecation,
    #[strum(
        message = "Fusion does not allow source overrides, so packages must disable a source explicitly instead."
    )]
    UnusedTables,
    #[strum(message = "Fusion reports this case under NoNodeForYamlKey instead.")]
    WrongResourceSchemaFile,
}
```

Author: Will add those that are missing!

### Warnings that are hard errors in Fusion

> Some <Constant name="core" /> warning names correspond to behaviors that <Constant name="fusion" /> enforces unconditionally as parse errors. If you reference these names in `warn_error_options`, <Constant name="fusion" /> emits a startup warning explaining that the entry has no effect. You can carry over your `warn_error_options` config from <Constant name="core" /> without breaking, but these configs do nothing:

Reviewer: > but these configs do nothing

as of today they will warn as "unsupported" ;-) we probably need to move them into "will not support" bucket to be clear with users and issue a warning suggesting they remove them from the config

> Some <Constant name="core" /> warning names correspond to behaviors that <Constant name="fusion" /> enforces unconditionally as parse errors. If you reference these names in `warn_error_options`, <Constant name="fusion" /> emits a startup warning explaining that the entry has no effect. You can carry over your `warn_error_options` config from <Constant name="core" /> without breaking, but these configs do nothing:

Author: ```suggestion
Some <Constant name="core" /> warning names correspond to behaviors that <Constant name="fusion" /> enforces unconditionally as parse errors. If you reference these names in `warn_error_options`, <Constant name="fusion" /> emits a startup warning explaining that the entry has no effect. You can carry over your `warn_error_options` config from <Constant name="core" /> without breaking, but these configs do nothing (They will throw a warning as `unsupported` and should be removed from the config):
```

| dbt-Core event name | Fusion behavior | Fusion error code |
|---|---|---|
| `DuplicateYAMLKeysDeprecation` | Fusion's YAML parser rejects duplicate keys as hard parse errors | `DuplicateConfigKey` (1059) |
| `CustomKeyInConfigDeprecation` | Unknown config keys are rejected via strict schema validation | `UnusedConfigKey` (1060) |
| `CustomTopLevelKeyDeprecation` | Unknown top-level schema keys are hard parse errors | `UnusedConfigKey` (1060) |
| `ResourceNamesWithSpacesDeprecation` | Resource names with spaces are rejected during name validation | `SchemaError` |
| `InvalidValueForField` | Field value failures are surfaced as hard parse errors via deserialization | `SerializationError` |
| `GenericJSONSchemaValidationDeprecation` | JSON schema validation failures are hard parse errors | `SerializationError` |
| `DuplicateNameDistinctNodeTypesDeprecation` | Caught as a hard error during node resolution | `SchemaError` |

### The `StaticAnalysis` warning group

> When running in `baseline` mode, <Constant name="fusion" />'s SQL comprehension engine emits static analysis diagnostics (type errors, unresolved references, implicit coercions, etc.) as warnings. These are grouped under the `StaticAnalysis` category, which maps to all <Constant name="fusion" /> frontend warning codes below 1000.

Reviewer: > These are grouped under the `StaticAnalysis` category, which maps to all <Constant name="fusion" /> frontend warning codes below 1000.

not implemented yet. so silencing them today would require specifying on a per-code basis

Author: Going to remove this section until we have support

> Some <Constant name="core" /> warning names correspond to behaviors that <Constant name="fusion" /> enforces unconditionally as parse errors. If you reference these names in `warn_error_options`, <Constant name="fusion" /> emits a startup warning explaining that the entry has no effect. You can carry over your `warn_error_options` config from <Constant name="core" /> without breaking, but these configs do nothing (They will throw a warning as `unsupported` and should be removed from the config):

Reviewer 2: ```suggestion
Some <Constant name="core" /> warning names correspond to behaviors that <Constant name="fusion" /> enforces unconditionally as parse errors. If you reference these names in `warn_error_options`, <Constant name="fusion" /> emits a startup warning explaining that the entry has no effect. You can carry over your `warn_error_options` configuration from <Constant name="core" /> without breaking your project, but these entries have no effect: <Constant name="fusion" /> still emits a startup warning that they are unsupported, and you should remove them from your configuration.:
```

Reviewer 2: Hiya Matt,

Just amended this slight to align the style guide recommendation of clear, neutral technical prose.

> | dbt-Core event name | Fusion behavior | Fusion error code |

Reviewer 2: Same as above, "dbt-Core" be "dbt-Core", "dbt-core" or have a constant variable?

> | dbt-Core event name | Fusion behavior | Fusion error code |

Reviewer 2: ```suggestion
| dbt-Core event name | <Constant name="fusion" />  behavior | <Constant name="fusion" />  error code |
```

|---|---|---|

> | `DuplicateYAMLKeysDeprecation` | Fusion's YAML parser rejects duplicate keys as hard parse errors | `DuplicateConfigKey` (1059) |

Reviewer 2: ```suggestion
| `DuplicateYAMLKeysDeprecation` | <Constant name="fusion" />'s YAML parser rejects duplicate keys as hard parse errors | `DuplicateConfigKey` (1059) |
```

flags:
  warn_error_options:
    error:
      - StaticAnalysis
```

Or use a numeric code range to target specific static analysis diagnostic types. See [Configuring `static_analysis`](/docs/fusion/new-concepts#configuring-static_analysis) for more on how static analysis modes work.

:::caution Enabling `--warn-error` with static analysis in `baseline` mode
If your project emits static analysis warnings and you use `--warn-error` (which promotes all warnings to errors), your project may fail unexpectedly. We recommend explicitly listing the warning categories you want to enforce rather than using `error: all` when `baseline` mode is active.
:::

> ### Deprecated `include` and `exclude` keys

Reviewer: > The legacy `include` and `exclude` fields for `warn_error_options` (deprecated in dbt Core v1.8) are not supported in <Constant name="fusion" />

not true. they are supported and used! Yes, we issue the `WEOIncludeExcludeDeprecation`, but that doesn't mean we discard them

> The legacy `include` and `exclude` fields for `warn_error_options` (deprecated in dbt Core v1.8) are not supported in <Constant name="fusion" />. If you use them, <Constant name="fusion" /> emits a `WEOIncludeExcludeDeprecation` warning (code 1086) and ignores the deprecated keys. Migrate to `error`, `warn`, and `silence` instead:

Author: ```suggestion
The legacy `include` and `exclude` fields for `warn_error_options` were deprecated in dbt Core v1.8 but are still supported in <Constant name="fusion" />. If you use them, <Constant name="fusion" /> emits a `WEOIncludeExcludeDeprecation` warning (code 1086) and ignores the deprecated keys. Migrate to `error`, `warn`, and `silence` instead:
```

> The legacy `include` and `exclude` fields for `warn_error_options` were deprecated in dbt Core v1.8 but are still supported in <Constant name="fusion" />. If you use them, <Constant name="fusion" /> emits a `WEOIncludeExcludeDeprecation` warning (code 1086) and ignores the deprecated keys. Migrate to `error`, `warn`, and `silence` instead:

Reviewer 2: ```suggestion
The legacy `include` and `exclude` fields for `warn_error_options` were deprecated in <Constant name="core" />v1.8 but are still supported in <Constant name="fusion" />. If you use them, <Constant name="fusion" /> emits a `WEOIncludeExcludeDeprecation` warning (code 1086) and ignores the deprecated keys. Migrate to `error`, `warn`, and `silence` instead:
```

```yaml

> # Before (Core ≤1.7)

Reviewer 2: should this be dbt Core?

flags:
  warn_error_options:
    include: all
    exclude:
      - NoNodesForSelectionCriteria

> # After (Core ≥1.8 and Fusion)

Reviewer 2: same as above

:::

<VersionBlock firstVersion="2.0">

## Fusion behavior and warning codes

The <Constant name="fusion_engine" /> fully supports `warn_error_options`. This section describes important differences from <Constant name="core" /> behavior.

Existing dbt-core event names fall into three categories:

- **Supported** Mapped to similar <Constant name="fusion" /> warning and behave approximately the same.
- **Won't be supported:** Those that we deliberately decided to not ever support.
- **Not supported yet:** Parsed, but do nothing yet.

### Warning codes in Fusion

In <Constant name="fusion" />, every warning has both a numeric code (for example, `1092`) and an event name (for example, `NoNodesForSelectionCriteria`). You can use either form interchangeably in `warn_error_options`, and both will appear in any warning messages printed at runtime:

```yaml
flags:
  warn_error_options:
    error:
      - NothingToDo    # by name
      - 1074           # by code (JinjaLogWarning / exceptions.warn())
    silence:
      - FreshnessConfigProblem   # by name
```

> Any value that is not a recognized numeric code, supported legacy event name, or supported group (`all`, `*`) causes <Constant name="fusion" /> to exit with an error at startup.

Reviewer: now this list includes fusion native names
